pkg/meta/distributed_lock.go

```diff
@@ -0,0 +1,316 @@
+	backoff time.Duration
+	lease   time.Duration
+}
+
+// NewDistributedLockBuilder creates a new DistributedLockBuilder.
+func NewDistributedLockBuilder() *DistributedLockBuilder {
+	return &DistributedLockBuilder{
+		backoff: 3 * time.Second,
+		lease:   30 * time.Second,
+	}
+}
+
+// SetBackoff sets the backoff time between two lock operations.
+// Default is 3 * time.Second.
+func (b *DistributedLockBuilder) SetBackoff(backoff time.Duration) *DistributedLockBuilder {
+	b.backoff = backoff
+	return b
+}
+
+// SetLease sets the lease time of the lock.
+// Default is 30 * time.Second.
+func (b *DistributedLockBuilder) SetLease(lease time.Duration) *DistributedLockBuilder {
+	b.lease = lease
+	return b
+}
+
+// Build builds a DistributedLock.
+func (b *DistributedLockBuilder) Build(
+	ctx context.Context, store DataStore, instanceID string, lockName string) *DistributedLock {
+	ctx = logutil.WithCategory(ctx, "distributed-lock")
+	return &DistributedLock{
+		outerCtx:   ctx,
+		store:      store,
+		instanceID: instanceID,
+
+		lockName: []byte(lockName),
+		backoff:  b.backoff,
+		lease:    b.lease,
+	}
+}
+
+func (d *DistributedLock) renewLockLoop(ctx context.Context) {
+	ticker := time.NewTicker(d.lease / 3)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			err := d.renew(ctx)
+			if err != nil {
+				logutil.BgLogger().Warn("renew lock failed", zap.Error(err))
+				return
+			}
+		}
+	}
+}
+
+func (d *DistributedLock) renew(ctx context.Context) error {
+	return d.getLockInTxnWithRetry(ctx,
+		func(lc *lockContent, txn DataTxn) (storeErr, otherErr error) {
+			if lc == nil || lc.InstanceID != d.instanceID {
+				logutil.Logger(d.outerCtx).Warn("lock is unexpectedly removed", zap.String("instanceID", d.instanceID))
+				return nil, nil
+			}
+
+			// Lock is exist, update it.
+			lc.StartTS = txn.StartTS()
+			data, err := lc.encode()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			err = txn.Set(d.lockName, data)
+			if err != nil {
+				return errors.Trace(err), nil
+			}
+			return nil, nil
+		})
+}
+
+var lockHeldErr error = errors.Errorf("lock is held")
+
+// Lock locks the distributed lock.
+func (d *DistributedLock) Lock() error {
+	newLockCreated := false
+	err := d.getLockInTxnWithRetry(d.outerCtx,
+		func(lc *lockContent, txn DataTxn) (storeErr, otherErr error) {
+			startTS := txn.StartTS()
+			// Lock is not exist or expired, create or overwrite it.
+			if lc == nil || d.lockExpired(lc, startTS) {
+				lc := newLockContent(startTS, d.instanceID)
+				data, err := lc.encode()
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				err = txn.Set(d.lockName, data)
+				if err != nil {
+					return errors.Trace(err), nil
+				}
+				newLockCreated = true
+				return nil, nil
+			}
+			// Lock is valid, wait for a while.
+			return lockHeldErr, nil
+		})
+	if err != nil {
+		return err
+	}
+	if !intest.InTest && newLockCreated {
```

Comment:
newLockCreated is always true here?

Reply:
No, `getLockInTxnWithRetry()` may encounter error without running `newLockCreated = true`. For example, the outer context is cancelled, or there is a decoding error on `lockContent`.